### DISCUSSION TITLE
关联对象 AssociatedObject 完全解析 · /ao

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#287] [Announcements]
https://draveness.me/ao

**Comment by zrx89757** [comment]
语言简洁，思路清晰；非常感谢楼主的总结分享！有一个小问题稍有疑惑，本人不熟悉C++的机制，只是勉强能看懂的水平。你在说AssociationsManager的时候，提到“它维护了 spinlock_t 和 AssociationsHashMap 的单例，初始化它的时候会调用 lock.lock() 方法，在析构时会调用 lock.unlock()”；如果初始化调用自旋锁，到析构的时候才释放锁的话；自旋锁不是持续请求锁，不会挂起线程么；那这样会不会太消耗性能了？？ 是我理解错了，还是你的笔误？如果看到，还望不吝赐教。

**Comment by leejunhui** [comment]
灯塔老哥的文章就是这么行云流水~

**Comment by draveness** [comment]
> 灯塔老哥的文章就是这么行云流水~

竟然还有人看...

**Comment by luckyfrank** [comment]
老哥写的好，厉害厉害！

**Comment by Kalmaegi** [comment]
> @draveness
> > 灯塔老哥的文章就是这么行云流水~
> 
> 竟然还有人看...

肯定的!

**Comment by James-Chi-Jie** [comment]
好棒

**Comment by wxm5558** [comment]
> @zrx89757
> 语言简洁，思路清晰；非常感谢楼主的总结分享！有一个小问题稍有疑惑，本人不熟悉C++的机制，只是勉强能看懂的水平。你在说AssociationsManager的时候，提到“它维护了 spinlock_t 和 AssociationsHashMap 的单例，初始化它的时候会调用 lock.lock() 方法，在析构时会调用 lock.unlock()”；如果初始化调用自旋锁，到析构的时候才释放锁的话；自旋锁不是持续请求锁，不会挂起线程么；那这样会不会太消耗性能了？？ 是我理解错了，还是你的笔误？如果看到，还望不吝赐教。

manager是个临时变量，每次调用setAssocia或getAssocia函数的时候，临时变量先调用构造函数（lock），set和get函数执行结束了manager回收调用析构函数（unlock），使用过程中所有manager类共享一个static的hashmap
